diff --git a/src/internal/__popcountdi2.c b/src/internal/__popcountdi2.c
new file mode 100644
index 00000000..b4b374a1
--- /dev/null
+++ b/src/internal/__popcountdi2.c
@@ -0,0 +1,16 @@
+
+__attribute__((__weak__))
+int __popcountdi2(unsigned long a)
+{
+    unsigned long nbits = 0;
+
+    /* Count the number of bits that are set */
+    for (unsigned long i = 0; i < 64; i++)
+    {
+        if ((a & (1LU << i)))
+            nbits++;
+    }
+
+    /* Return 1 if the nbits is odd; return 0 if nbits is event */
+    return (nbits % 2) ? 1 : 0;
+}
diff --git a/src/stdio/__fprintf_chk.c b/src/stdio/__fprintf_chk.c
new file mode 100644
index 00000000..50253c11
--- /dev/null
+++ b/src/stdio/__fprintf_chk.c
@@ -0,0 +1,23 @@
+#include <stdio.h>
+#include <stdarg.h>
+
+extern int __vfprintf_chk(
+    FILE *stream,
+    int flag,
+    const char *format,
+    va_list ap);
+
+__attribute__((__weak__))
+int __fprintf_chk(FILE *stream, int flag, const char *format, ...)
+{
+    int ret;
+    va_list ap;
+
+    (void)flag;
+
+    va_start(ap, format);
+    ret = __vfprintf_chk(stream, flag, format, ap);
+    va_end(ap);
+
+    return ret;
+}
diff --git a/src/stdio/__vfprintf_chk.c b/src/stdio/__vfprintf_chk.c
new file mode 100644
index 00000000..5493e453
--- /dev/null
+++ b/src/stdio/__vfprintf_chk.c
@@ -0,0 +1,11 @@
+#include <stdio.h>
+#include <assert.h>
+
+__attribute__((__weak__))
+int __vfprintf_chk(FILE *stream, int flag, const char *format, va_list ap)
+{
+    assert(stream);
+    assert(format);
+    (void)flag;
+    return vfprintf(stream, format, ap);
+}
diff --git a/src/thread/__unmapself.c b/src/thread/__unmapself.c
index 31d94e67..fecb7ea5 100644
--- a/src/thread/__unmapself.c
+++ b/src/thread/__unmapself.c
@@ -6,12 +6,12 @@
 
 static void *unmap_base;
 static size_t unmap_size;
-static char shared_stack[256];
+static char shared_stack[4096];
 
 static void do_unmap()
 {
-	__syscall(SYS_munmap, unmap_base, unmap_size);
-	__syscall(SYS_exit);
+	const long SYS_myst_clean_exit = 1016;
+	__syscall(SYS_myst_clean_exit, 0, unmap_base, unmap_size);
 }
 
 void __unmapself(void *base, size_t size)
diff --git a/src/thread/x86_64/__unmapself.s b/src/thread/x86_64/__unmapself.s
deleted file mode 100644
index e2689e65..00000000
--- a/src/thread/x86_64/__unmapself.s
+++ /dev/null
@@ -1,10 +0,0 @@
-/* Copyright 2011-2012 Nicholas J. Kain, licensed under standard MIT license */
-.text
-.global __unmapself
-.type   __unmapself,@function
-__unmapself:
-	movl $11,%eax   /* SYS_munmap */
-	syscall         /* munmap(arg2,arg3) */
-	xor %rdi,%rdi   /* exit() args: always return success */
-	movl $60,%eax   /* SYS_exit */
-	syscall         /* exit(0) */
